contributors/design-proposals/customresources-subresources.md

```diff
@@ -0,0 +1,221 @@
+2. Support a `/scale` subresource for CustomResources.
+3. Maintain backward compatibility by allowing CRDs to opt-in to enable subresources.
+4. If a CustomResource is already structured using spec/status, allow it to easily transition to use the `/status` and `/scale` endpoint.
+5. Work seamlessly with [JSON Schema validation](https://github.com/kubernetes/community/pull/708).
+
+## Non-Goals
+
+1. Allow defining arbitrary subresources i.e. subresources except `/status` and `/scale`.
+2. Unify the many `Scale` types: as long as there is no generic `Scale` object in Kubernetes, we will propose to introduce yet another `Scale` type in the `apiextensions.k8s.io` api group. If Kubernetes switches to a generic `Scale` object, `apiextensions.k8s.io` will follow.
+
+## Proposed Extension of CustomResourceDefinition
+
+### API Types
+
+The addition of the following external types in `apiextensions.k8s.io/v1beta1` is proposed:
+
+```go
+type CustomResourceDefinitionSpec struct {
+    ...
+    // SubResources describes the subresources for CustomResources
+    // This field is alpha-level and should only be sent to servers that enable
+    // subresources via the CurstomResourceSubResources feature gate.
+    // +optional
+    SubResources *CustomResourceSubResources `json:“subResources,omitempty”`
+}
+
+// CustomResourceSubResources defines the status and scale subresources for CustomResources.
+type CustomResourceSubResources struct {
+    // Status denotes the status subresource for CustomResources
+    Status *CustomResourceSubResourceStatus `json:“status,omitempty”`
+    // Scale denotes the scale subresource for CustomResources
+    Scale  *CustomResourceSubResourceScale  `json:“scale,omitempty”`
+}
+
+// CustomResourceSubResourceStatus defines how to serve the HTTP path <CR Name>/status.
+type CustomResourceSubResourceStatus struct {
+    // The JSON path (e.g. “.status”) of the status of a CustomResource.
+    // The whole object is transferred over the wire, but only the status can be mutated via the /status subresource.
+    // StatusPath is restricted to be “.status” and defaults to “.status”.
+    StatusPath string `json:“statusPath,omitempty”`
+    // The JSON path (e.g. “.spec”) of the spec of a CustomResource.
+    // SpecPath is restricted to be “.spec” and defaults to “.spec”.
+    // Changes to the specified JSON path increase the “.metadata.generation” value.
+    SpecPath string `json:“specPath,omitempty”`
+}
+
+// CustomResourceSubResourceScale defines how to serve the HTTP path <CR name>/scale.
+type CustomResourceSubResourceScale struct {
+    // required, e.g. “.spec.replicas”.
+    // Only JSON paths without the array notation are allowed.
+    SpecReplicasPath string `json:“specReplicasPath,omitempty”`
+    // optional, e.g. “.status.replicas”.
+    // Only JSON paths without the array notation are allowed.
+    StatusReplicasPath string `json:“statusReplicasPath,omitempty”`
```

Comment:
My argument was that the purpose of the /scale subresource, as far as I know, is to abstract away the knowledge of which field a given object considers to represent its "scale". If we wanted to go the route of a convention for field names, like it must be `.status.replicas`, then there would be no need for a subresource. You could just look directly at `.status.replicas`.

This is in contrast to the `/status` subresource, which I argued ought to be restricted to the convention of `.spec` and `.status` field names. The `/status` endpoint returns the whole object, not a sub-field or a new type like `/scale` does, and so the only way the caller knows which part is the status is by agreeing on the conventional field name. Similarly, the caller needs to agree that `.spec` is the field name that triggers `.metadata.generation` to be incremented and so on.

Because `/scale` instead returns a separate, abstracted type, there is no need to agree on the underlying field names.